### PR TITLE
Bug/jenkins 47752 jdl corejs lint tests enable

### PR DESCRIPTION
**Pull request by cliffmeyers**
# Description
- See [JENKINS-47752](https://issues.jenkins-ci.org/browse/JENKINS-47752).
- Status: JDL and core-js should both have lint and tests being run properly. There are two outstanding issues:
   - [x] Figure out why Activity-spec from dashboard is broken
   - [x] Figure out why JS test failures are not being recorded in CI.

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate.
- [x] Change is code complete and matches issue description
- [x] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

### CODE REVIEW

**Comment by cliffmeyers** [comment]
Lint and tests should now be enabled for JDL and core-js. Unfortunately it looks like tests are failing in dashboard (possibly other modules too) because it can't import some things from blueocean-core-js.

```
[ERROR] Summary of all failing tests
[ERROR]  FAIL  src/test/js/pipeline-graph-converter-spec.js
[ERROR]   ● Test suite failed to run
[ERROR] 
[ERROR]     Cannot find module '@jenkins-cd/blueocean-core-js' from 'PipelineRunGraph.jsx'
[ERROR]       
[ERROR]       at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:169:17)
[ERROR]       at Object.<anonymous> (src/main/js/components/PipelineRunGraph.jsx:96:315)
[ERROR] 
[ERROR]  FAIL  src/test/js/push-events-actions-spec.js
[ERROR]   ● Test suite failed to run
[ERROR] 
[ERROR]     Cannot find module '@jenkins-cd/blueocean-core-js' from 'smart-fetch.js'
[ERROR]       
[ERROR]       at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:169:17)
[ERROR]       at Object.<anonymous> (src/main/js/util/smart-fetch.js:193:225)
[ERROR] 
[ERROR]  FAIL  src/test/js/testResult-spec.js
[ERROR]   ● Test suite failed to run
[ERROR] 
[ERROR]     Cannot find module '@jenkins-cd/blueocean-core-js' from 'TestSection.jsx'
[ERROR]       
[ERROR]       at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:169:17)
[ERROR]       at Object.<anonymous> (src/main/js/components/testing/TestSection.jsx:3:50)
[ERROR] 
[ERROR]  FAIL  src/test/js/UrlUtils-spec.js
[ERROR]   ● Test suite failed to run
[ERROR] 
[ERROR]     Cannot find module '@jenkins-cd/blueocean-core-js' from 'UrlUtils-spec.js'
[ERROR]       
[ERROR]       at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:169:17)
[ERROR]       at Object.<anonymous> (src/test/js/UrlUtils-spec.js:3:24)
[ERROR] 
[ERROR]  FAIL  src/test/js/activity-spec.js
[ERROR]   ● Test suite failed to run
[ERROR] 
[ERROR]     Cannot find module '@jenkins-cd/blueocean-core-js' from 'Activity.jsx'
[ERROR]       
[ERROR]       at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:169:17)
[ERROR]       at Object.<anonymous> (src/main/js/components/Activity.jsx:8:50)
[ERROR] 
[ERROR]  FAIL  src/test/js/log-spec.js
[ERROR]   ● Test suite failed to run
[ERROR] 
[ERROR]     Cannot find module '@jenkins-cd/blueocean-core-js' from 'log-spec.js'
[ERROR]       
[ERROR]       at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:169:17)
[ERROR]       at Object.<anonymous> (src/test/js/log-spec.js:6:24)
[ERROR] 
[ERROR]  FAIL  src/test/js/store-spec.js
[ERROR]   ● Test suite failed to run
[ERROR] 
[ERROR]     Cannot find module '@jenkins-cd/blueocean-core-js' from 'smart-fetch.js'
[ERROR]       
[ERROR]       at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:169:17)
[ERROR]       at Object.<anonymous> (src/main/js/util/smart-fetch.js:193:225)
[ERROR] 
[ERROR]  FAIL  src/test/js/pipeline-spec.js
[ERROR]   ● Test suite failed to run
[ERROR] 
[ERROR]     Cannot find module '@jenkins-cd/blueocean-core-js' from 'UrlUtils.js'
[ERROR]       
[ERROR]       at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:169:17)
[ERROR]       at Object.<anonymous> (src/main/js/util/UrlUtils.js:1:1031)
[ERROR] 
[ERROR]  FAIL  src/test/js/smart-fetch-spec.js
[ERROR]   ● Test suite failed to run
[ERROR] 
[ERROR]     Cannot find module '@jenkins-cd/blueocean-core-js' from 'smart-fetch.js'
[ERROR]       
[ERROR]       at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:169:17)
[ERROR]       at Object.<anonymous> (src/main/js/util/smart-fetch.js:193:225)
[ERROR] 
[ERROR]  FAIL  src/test/js/store-rundetails-spec.js
[ERROR]   ● Test suite failed to run
[ERROR] 
[ERROR]     Cannot find module '@jenkins-cd/blueocean-core-js' from 'logDisplayHelper.js'
[ERROR]       
[ERROR]       at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:169:17)
[ERROR]       at Object.<anonymous> (src/main/js/util/logDisplayHelper.js:2:50)
[ERROR] 
[ERROR]  FAIL  src/test/js/pipelines-spec.js
[ERROR]   ● Test suite failed to run
[ERROR] 
[ERROR]     Cannot find module '@jenkins-cd/blueocean-core-js' from 'Pipelines.jsx'
[ERROR]       
[ERROR]       at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:169:17)
[ERROR]       at Object.<anonymous> (src/main/js/components/Pipelines.jsx:4:50)
[ERROR] 
[ERROR]  FAIL  src/test/js/LogConsole-spec.js
[ERROR]   ● Test suite failed to run
[ERROR] 
[ERROR]     Cannot find module '@jenkins-cd/blueocean-core-js' from 'LogConsole.jsx'
[ERROR]       
[ERROR]       at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:169:17)
[ERROR]       at Object.<anonymous> (src/main/js/components/karaoke/components/LogConsole.jsx:3:50)
[ERROR] 
[ERROR]  FAIL  src/test/js/run-details-changes-spec.js
[ERROR]   ● Test suite failed to run
[ERROR] 
[ERROR]     Cannot find module '@jenkins-cd/blueocean-core-js' from 'run-details-changes-spec.js'
[ERROR]       
[ERROR]       at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:169:17)
[ERROR]       at Object.<anonymous> (src/test/js/run-details-changes-spec.js:4:24)
[ERROR] 
[ERROR]  FAIL  src/test/js/run-details-artifacts-spec.js
[ERROR]   ● Test suite failed to run
[ERROR] 
[ERROR]     Cannot find module '@jenkins-cd/blueocean-core-js' from 'RunDetailsArtifacts.jsx'
[ERROR]       
[ERROR]       at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:169:17)
[ERROR]       at Object.<anonymous> (src/main/js/components/RunDetailsArtifacts.jsx:6:50)
[ERROR] 
[ERROR]  FAIL  src/test/js/pipeline-not-found-spec.js
[ERROR]   ● Test suite failed to run
[ERROR] 
[ERROR]     Cannot find module '@jenkins-cd/blueocean-core-js' from 'PipelinePage.jsx'
[ERROR]       
[ERROR]       at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:169:17)
[ERROR]       at Object.<anonymous> (src/main/js/components/PipelinePage.jsx:10:50)
[ERROR] 
[ERROR]  FAIL  src/test/js/pullRequests-spec.js
[ERROR]   ● Test suite failed to run
[ERROR] 
[ERROR]     Cannot find module '@jenkins-cd/blueocean-core-js' from 'PullRequests.jsx'
[ERROR]       
[ERROR]       at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:169:17)
[ERROR]       at Object.<anonymous> (src/main/js/components/PullRequests.jsx:3:50)
[ERROR] 
[ERROR]  FAIL  src/test/js/creation/link-spec.js
[ERROR]   ● Test suite failed to run
[ERROR] 
[ERROR]     Cannot find module '@jenkins-cd/blueocean-core-js' from 'ClassicCreationLink.jsx'
[ERROR]       
[ERROR]       at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:169:17)
[ERROR]       at Object.<anonymous> (src/main/js/creation/ClassicCreationLink.jsx:8:220)
[ERROR] 
[ERROR]  FAIL  src/test/js/creation/git/GitFlowManager-spec.js
[ERROR]   ● Test suite failed to run
[ERROR] 
[ERROR]     Cannot find module '@jenkins-cd/blueocean-core-js' from 'GitFlowManager.jsx'
[ERROR]       
[ERROR]       at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:169:17)
[ERROR]       at Object.<anonymous> (src/main/js/creation/git/GitFlowManager.jsx:5:50)
[ERROR] 
[ERROR] 
[ERROR] Test Suites: 18 failed, 11 passed, 29 total
[ERROR] Tests:       69 passed, 69 total
[ERROR] Snapshots:   0 total
[ERROR] Time:        40.373s
[ERROR] Ran all test suites.
[INFO] [15:52:42] tests failed or errored!
```

cc @kzantow I am guessing that we aren't doing something in CI build to push the contents of core-js into node_modules for dashboard and other modules that depend on core-js. Do you know of a quick fix for this?

**Comment by cliffmeyers** [comment]
This branch is also affected by the issue that @michaelneale saw where JS test failures are not being reported correctly to mark the build as unstable.

**Comment by cliffmeyers** [comment]
Running locally, all tests for `blueocean-dashboard` pass except for Activity-spec:

```
[ERROR] Summary of all failing tests
[ERROR]  FAIL  src/test/js/activity-spec.js (8.626s)
[ERROR]   ● Pipeline -> Activity List › should contain cause
[ERROR] 
[ERROR]     TypeError: _moment2.default.duration(...).format is not a function
[ERROR]       
[ERROR]       at Function.format (node_modules/@jenkins-cd/design-language/dist/js/components/TimeDuration.js:160:57)
[ERROR]       at TimeDuration.render (node_modules/@jenkins-cd/design-language/dist/js/components/TimeDuration.js:130:45)
[ERROR]       at ReactCompositeComponent._renderValidatedComponentWithoutOwnerOrContext (node_modules/react-dom/lib/ReactCompositeComponent.js:799:30)
[ERROR]       at ReactCompositeComponent._renderValidatedComponent (node_modules/react-dom/lib/ReactCompositeComponent.js:822:32)
[ERROR]       at ReactCompositeComponent.performInitialMount (node_modules/react-dom/lib/ReactCompositeComponent.js:362:30)
[ERROR]       at ReactCompositeComponent.mountComponent (node_modules/react-dom/lib/ReactCompositeComponent.js:258:21)
[ERROR]       at Object.ReactReconciler.mountComponent (node_modules/react-dom/lib/ReactReconciler.js:46:35)
[ERROR]       at ReactDOMComponent.ReactMultiChild.Mixin.mountChildren (node_modules/react-dom/lib/ReactMultiChild.js:238:44)
[ERROR]       at ReactDOMComponent.Mixin._createContentMarkup (node_modules/react-dom/lib/ReactDOMComponent.js:653:32)
[ERROR]       at ReactDOMComponent.Mixin.mountComponent (node_modules/react-dom/lib/ReactDOMComponent.js:520:29)
[ERROR]       at Object.ReactReconciler.mountComponent (node_modules/react-dom/lib/ReactReconciler.js:46:35)
[ERROR]       at ReactDOMComponent.ReactMultiChild.Mixin.mountChildren (node_modules/react-dom/lib/ReactMultiChild.js:238:44)
[ERROR]       at ReactDOMComponent.Mixin._createContentMarkup (node_modules/react-dom/lib/ReactDOMComponent.js:653:32)
[ERROR]       at ReactDOMComponent.Mixin.mountComponent (node_modules/react-dom/lib/ReactDOMComponent.js:520:29)
[ERROR]       at Object.ReactReconciler.mountComponent (node_modules/react-dom/lib/ReactReconciler.js:46:35)
[ERROR]       at ReactCompositeComponent.performInitialMount (node_modules/react-dom/lib/ReactCompositeComponent.js:371:34)
[ERROR]       at ReactCompositeComponent.mountComponent (node_modules/react-dom/lib/ReactCompositeComponent.js:258:21)
[ERROR]       at Object.ReactReconciler.mountComponent (node_modules/react-dom/lib/ReactReconciler.js:46:35)
[ERROR]       at ReactCompositeComponent.performInitialMount (node_modules/react-dom/lib/ReactCompositeComponent.js:371:34)
[ERROR]       at ReactCompositeComponent.mountComponent (node_modules/react-dom/lib/ReactCompositeComponent.js:258:21)
[ERROR]       at Object.ReactReconciler.mountComponent (node_modules/react-dom/lib/ReactReconciler.js:46:35)
[ERROR]       at ReactDOMComponent.ReactMultiChild.Mixin.mountChildren (node_modules/react-dom/lib/ReactMultiChild.js:238:44)
[ERROR]       at ReactDOMComponent.Mixin._createContentMarkup (node_modules/react-dom/lib/ReactDOMComponent.js:653:32)
[ERROR]       at ReactDOMComponent.Mixin.mountComponent (node_modules/react-dom/lib/ReactDOMComponent.js:520:29)
[ERROR]       at Object.ReactReconciler.mountComponent (node_modules/react-dom/lib/ReactReconciler.js:46:35)
[ERROR]       at ReactCompositeComponent.performInitialMount (node_modules/react-dom/lib/ReactCompositeComponent.js:371:34)
[ERROR]       at ReactCompositeComponent.mountComponent (node_modules/react-dom/lib/ReactCompositeComponent.js:258:21)
[ERROR]       at Object.ReactReconciler.mountComponent (node_modules/react-dom/lib/ReactReconciler.js:46:35)
[ERROR]       at ReactCompositeComponent.performInitialMount (node_modules/react-dom/lib/ReactCompositeComponent.js:371:34)
[ERROR]       at ReactCompositeComponent.mountComponent (node_modules/react-dom/lib/ReactCompositeComponent.js:258:21)
[ERROR]       at Object.ReactReconciler.mountComponent (node_modules/react-dom/lib/ReactReconciler.js:46:35)
[ERROR]       at ReactCompositeComponent.performInitialMount (node_modules/react-dom/lib/ReactCompositeComponent.js:371:34)
[ERROR]       at ReactCompositeComponent.mountComponent (node_modules/react-dom/lib/ReactCompositeComponent.js:258:21)
[ERROR]       at Object.ReactReconciler.mountComponent (node_modules/react-dom/lib/ReactReconciler.js:46:35)
[ERROR]       at node_modules/react-dom/lib/ReactServerRendering.js:45:36
[ERROR]       at ReactServerRenderingTransaction.TransactionImpl.perform (node_modules/react-dom/lib/Transaction.js:140:20)
[ERROR]       at renderToStringImpl (node_modules/react-dom/lib/ReactServerRendering.js:43:24)
[ERROR]       at renderToStaticMarkup (node_modules/react-dom/lib/ReactServerRendering.js:83:10)
[ERROR]       at ShallowWrapper.<anonymous> (node_modules/enzyme/build/ShallowWrapper.js:701:72)
[ERROR]       at ShallowWrapper.single (node_modules/enzyme/build/ShallowWrapper.js:1398:19)
[ERROR]       at ShallowWrapper.html (node_modules/enzyme/build/ShallowWrapper.js:698:21)
[ERROR]       at Object.<anonymous> (src/test/js/activity-spec.js:202:27)
[ERROR] 
[ERROR] 
[ERROR] Test Suites: 1 failed, 2 skipped, 26 passed, 27 of 29 total
[ERROR] Tests:       1 failed, 8 skipped, 149 passed, 158 total
[ERROR] Snapshots:   0 total
[ERROR] Time:        27.526s
[ERROR] Ran all test suites.
[INFO] [12:36:49] tests failed or errored!
```

File: blueocean-core-js/package.json [hunk 1]
```diff
@@ -76,15 +76,10 @@
     "gulp-babel": "6.1.2",
     "gulp-copy": "0.0.2",
     "gulp-eslint": "3.0.1",
-    "gulp-jest": "1.0.0",
     "gulp-less": "3.1.0",
     "gulp-rename": "1.2.2",
     "gulp-sourcemaps": "1.6.0",
-    "jest": "19.0.2",
```
Comment (scherler) [inline blueocean-core-js/package.json]: I am tending to include those "second hand" libraries lately since not having them explicitly declared can lead to weird versions.

File: jenkins-design-language/gulpfile.js [hunk 2]
```diff
@@ -246,9 +183,20 @@ var builder = require('@jenkins-cd/js-builder');
 builder.src([
     'src/js',
     'less',
-    'dist' // for icons & fonts
+    'dist' // for icons & fonts; NOTE: would be nice to find another way to do this as files in dist creates issues for jest & eslint
 ]);
 
+builder.tests('test/js');
```
Comment (scherler) [inline jenkins-design-language/gulpfile.js]: You have to define that because we do not use the "standard" location? If not since gulp-runner 1.1.0 you should not need this. We should fall back to the declared tasks in jsBuider

File: jenkins-design-language/package.json [hunk 3]
```diff
@@ -8,16 +8,15 @@
     "storybook": "npm run gulp build && start-storybook -p 9001 --static-dir dist/assets/",
     "build-storybook": "build-storybook --static-dir dist/assets/ --output-dir website/storybook",
     "flow": "flow",
+    "lint": "gulp lint",
     "gulp": "gulp build",
-    "test": "gulp test",
-    "prepublish": "gulp test",
```
Comment (scherler) [inline jenkins-design-language/package.json]: No prepublish anymore?

File: jenkins-design-language/src/js/components/Linkify.jsx [hunk 4]
```diff
@@ -1,6 +1,7 @@
 // @flow
 
 import React from 'react';
+import { PropTypes } from 'react';
```
Comment (scherler) [inline jenkins-design-language/src/js/components/Linkify.jsx]: Actually why do you not use https://www.npmjs.com/package/prop-types ? ...and why a second line for it?

File: jenkins-design-language/src/js/components/Linkify.jsx [hunk 5]
```diff
@@ -36,5 +37,8 @@ export const Linkify = (props: Props) => {
     );
 };
 
-Linkify.propTypes = OriginalLinkify.propTypes;
+Linkify.propTypes = {
+    ...OriginalLinkify.propTypes,
+    options: PropTypes.object,
```
Comment (scherler) [inline jenkins-design-language/src/js/components/Linkify.jsx]: I have the feeling this change is not related to the underlying PR

File: blueocean-core-js/package.json (hunk 1 above)
Comment (michaelneale) [reply blueocean-core-js/package.json]: yes this would explain a lot ;)

File: jenkins-design-language/package.json (hunk 3 above)
Comment (cliffmeyers) [reply jenkins-design-language/package.json]: Absolutely no prepublish. 1. We are not publishing stuff to npm anymore, so it's not useful, 2. prepublish is called anytime that "npm install" is called... so when frontend-maven-plugin invokes npm install, it runs the tests a 2nd time, which adds 1m+ to the build

File: jenkins-design-language/gulpfile.js (hunk 2 above)
Comment (cliffmeyers) [reply jenkins-design-language/gulpfile.js]: Yes, this is used because the project structure of JDL and core-js are subtly different than dashboard, personalization, etc.

File: blueocean-core-js/package.json (hunk 1 above)
Comment (cliffmeyers) [reply blueocean-core-js/package.json]: Yep, good tip, I may end up adding `jest` back in once I sort out some of these other problems.

File: jenkins-design-language/src/js/components/Linkify.jsx (hunk 4 above)
Comment (cliffmeyers) [reply jenkins-design-language/src/js/components/Linkify.jsx]: I don't think prop-types is available under our React version, and the reason I split it to a second line is because I know we'll need to change it to import from a different package in a later PR when we do the upgrade. ;)

File: jenkins-design-language/src/js/components/Linkify.jsx (hunk 5 above)
Comment (cliffmeyers) [reply jenkins-design-language/src/js/components/Linkify.jsx]: Because lint was not being run during the build, this lint error was not failing the build. Now that lint is being running, this lint error needs to be fixed in order for the build to pass.

**Comment by cliffmeyers** [comment]
I built our CI image locally and ran the build inside of it:
- The errors with missing blueocean-core-js imports is reproducible,
- It appears the contents of blueocean-core-js/dist/**/* is not being packaged into the .jar and .hpi file. This appears to be specific to Linux... still checking this out.

Other note, JDL and core-js use different parent POM's. I have to assume this is a mistake and will fix it.

**Comment by cliffmeyers** [comment]
I think the problem is that maven-resources-plugin fires before the gulp build does, so in clean environments the dist directory doesn't actually exist.

**Comment by michaelneale** [comment]
@kzantow does this have anything to do with what you are working on now?

**Comment by cliffmeyers** [comment]
My changes from yesterday seem to have restored the ability for test results to be recorded properly in CI. See these builds:

https://ci.blueocean.io/blue/organizations/jenkins/blueocean/detail/bug%2FJENKINS-47752-jdl-corejs-lint-tests-enable/7/tests

https://ci.blueocean.io/blue/organizations/jenkins/blueocean/detail/bug%2FJENKINS-47752-jdl-corejs-lint-tests-enable/8/tests

**Comment by michaelneale** [comment]
nice one @cliffmeyers 👍🏻

**Comment by vivek** [comment]
LGTM🐝 thanks @cliffmeyers!

**Comment by cliffmeyers** [comment]
Before changes:

screen shot 2017-10-31 at 10 51 04 am

After changes:

screen shot 2017-11-01 at 1 57 10 pm

Note that JS tests are now accounted for.

**Comment by cliffmeyers** [comment]
I created [JENKINS-47774](https://issues.jenkins-ci.org/browse/JENKINS-47774) to track the work of re-enabling the mystery failing test that I marked as skipped in 56433a1

File: jenkins-design-language/package.json (hunk 3 above)
Comment (scherler) [reply jenkins-design-language/package.json]: Actually not publishing anymore to npm means that third party integrators cannot use JDL anymore. e.g. it is impossible to test with storybook since the versions are not published anymore. :(

@cliffmeyers ...the whole move to "hpi as only truth" has completely made integration of bo/jdl components in third party components impossible! re-Usability = 0 :(

File: jenkins-design-language/src/js/components/Linkify.jsx (hunk 4 above)
Comment (scherler) [reply jenkins-design-language/src/js/components/Linkify.jsx]: That package is completely independent of the underlying react version.

File: jenkins-design-language/src/js/components/Linkify.jsx (hunk 5 above)
Comment (scherler) [reply jenkins-design-language/src/js/components/Linkify.jsx]: Yeah got you but then it should not have been in this PR but another PR extending this one, so this PR would show that CI should fail and the other then merged with this fix.

File: jenkins-design-language/package.json (hunk 3 above)
Comment (cliffmeyers) [reply jenkins-design-language/package.json]: Can you explain what you mean by "third party component" ?  If a Jenkins plugin depends on the JDL and is using `blueocean-maven-plugin` then it will receive the contents of the JDL's "dist" dir inside of `node_modules/@jenkins-cd/design-language`.

Comment (cliffmeyers) [reply jenkins-design-language/package.json]: Regardless, we can still publish the JDL periodically if something totally disconnected from Blue Ocean wants to use it. `prepublish` isn't really the right hook to be doing a `gulp test` given that `prepublish` runs any time a parameterless `npm install` is run.

File: jenkins-design-language/src/js/components/Linkify.jsx (hunk 4 above)
Comment (cliffmeyers) [reply jenkins-design-language/src/js/components/Linkify.jsx]: First version of `prop-types` package is 15.5.0 and we're on 15.4.0. Even though React is good about compatibility, I don't think mixing and matching versions is a good idea. I don't think splitting an import into two lines is really a big deal is it? :)

File: jenkins-design-language/package.json (hunk 3 above)
Comment (cliffmeyers) [reply jenkins-design-language/package.json]: We could start by publishing JDL and core-js whenever we do a minor or major release of Blue Ocean, e.g. 1.4 or 2.0. Would that help?

Comment (scherler) [reply jenkins-design-language/package.json]: IMO **could** not but **have to**, for third party is e.g. storybook a good example. You do only get the "correct" version **after** you have installed/started a jenkins, which IMO defeats the purpose of storybook. My point in using storybook to not have to start jenkins. Storybook should get their deps via a **normal** `npm i`.

File: jenkins-design-language/src/js/components/Linkify.jsx (hunk 4 above)
Comment (scherler) [reply jenkins-design-language/src/js/components/Linkify.jsx]: well I do not like having two lines and you are right they splitted the lib in 15.5 but that lib has no dep on any react, it is an independent util lib that you can even use in react from a year ago.

File: jenkins-design-language/package.json (hunk 3 above)
Comment (cliffmeyers) [reply jenkins-design-language/package.json]: You really just need to run a Maven target to get the right deps, such as `mvn process-sources`. I realize it's not ideal, but you could built that into npm scripts or write a shell script to it for you.

**Comment by michaelneale** [comment]
I think @kzantow was saying that it could also be published as an npm module

File: jenkins-design-language/package.json (hunk 3 above)
Comment (scherler) [reply jenkins-design-language/package.json]: So the verdict is `jdl` is a jenkins only component where you need full java installed to be able to use it.

...there is *no* need for that but I guess to do `npm publish` is just too much.

Comment (michaelneale) [reply jenkins-design-language/package.json]: No as stated elsewhere it could and probably should be published to npm cc @kzantow could we still publish it? (Separate PR)

Comment (kzantow) [reply jenkins-design-language/package.json]: Right @scherler we can still publish to npm, there's nothing preventing that. The only possible issue is versions causing some horrible annoyances with the `blueocean-maven-plugin`, but I could probably change that to write an appropriate version.
